### PR TITLE
Article feed block

### PR DESCRIPTION
**Pull request by ninoaguilar**
* Migrated article feed block from blog to milo. 
* Migrated over the `taxonomy.js` script
* Code refactor on migrated code.

Resolves: [GitHub - Article Feed Block Ticket](https://github.com/adobecom/blog/issues/4)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/naguilar/article-feed?martech=off
- After:  https://article-feed-block--milo--adobecom.hlx.page/drafts/naguilar/article-feed?martech=off

### CODE REVIEW

**Comment by auniverseaway** [comment]
@ninoaguilar thanks for the contribution! Lots of cool stuff in here!

The first thing that jumps out to me is that utils shouldn't have to change as much as you have done. loadTaxonomy, and some of the other functions are not a utils area of concern... 99% of Milo pages will not need taxonomy. Let's jump on a call on Monday to see how we can better isolate your needed features.

**Comment by ninoaguilar** [comment]
Sounds good @auniverseaway ! Yes, I would like to know where/when to create new files to better contain some of these features.

File: libs/utils/utils.js [hunk 1]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
+
+export const getLocaleIetf = () => getConfig().locale?.ietf;
+
+/**
+ * Returns the language dependent root path
+ * @returns {string} The computed root path
+ */
+export const getPrefix = () => getConfig().locale?.prefix;
+
+export const emptyDiv = () => document.createElement('div');
```
Comment (auniverseaway) [inline libs/utils/utils.js:513]: Use `createTag` instead.

File: libs/utils/utils.js [hunk 2]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
   return io;
 }
+
+export const getLocaleIetf = () => getConfig().locale?.ietf;
+
+/**
+ * Returns the language dependent root path
+ * @returns {string} The computed root path
+ */
+export const getPrefix = () => getConfig().locale?.prefix;
```
Comment (auniverseaway) [inline libs/utils/utils.js:511]: Not helpful enough to warrant an abstraction being in utils.

File: libs/utils/utils.js [hunk 3]
```diff
@@ -91,12 +94,12 @@ function getEnv(conf) {
...
 export function getLocale(locales) {
   if (!locales) {
-    return { ietf: 'en-US', tk: 'hah7vzn.css', prefix: '' };
+    return { ietf: 'en-US', tk: 'hah7vzn.css', prefix: 'en' };
   }
   const { pathname } = window.location;
   const split = pathname.split('/');
   const locale = locales[split[1]] || locales[''];
-  locale.prefix = locale.ietf === 'en-US' ? '' : `/${split[1]}`;
+  locale.prefix = locale.ietf === 'en-US' ? 'en' : `/${split[1]}`;
```
Comment (auniverseaway) [inline libs/utils/utils.js:102]: This needs to be changed back. There's reasons we have it a certain way. I can work with you on the blog use case.

File: libs/utils/utils.js [hunk 4]
```diff
@@ -91,12 +94,12 @@ function getEnv(conf) {
 // find out current locale based on pathname and existing locales object from config.
 export function getLocale(locales) {
   if (!locales) {
-    return { ietf: 'en-US', tk: 'hah7vzn.css', prefix: '' };
+    return { ietf: 'en-US', tk: 'hah7vzn.css', prefix: 'en' };
```
Comment (auniverseaway) [inline libs/utils/utils.js:97]: This needs to be changed back. There's reasons we have it a certain way. I can work with you on the blog use case.

File: libs/utils/utils.js [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import * as taxonomyLibrary from '../scripts/taxonomy.js';
```
Comment (auniverseaway) [inline libs/utils/utils.js:1]: This shouldn't be here. Only blocks that need this should be importing it. An entire site should not have to pay the performance penalty of this import.

File: libs/scripts/taxonomy.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
```
Comment (auniverseaway) [inline libs/scripts/taxonomy.js:2]: 2022

File: libs/utils/utils.js [hunk 7]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
+    const json = await resp.json();
+    window.placeholders = {};
+    json.data.forEach((placeholder) => {
+      window.placeholders[placeholder.Key] = placeholder.Text;
+    });
+  }
+  return window.placeholders;
+}
+
+export function stamp(message) {
```
Comment (auniverseaway) [inline libs/utils/utils.js:532]: I'd remove until we decide as a group we want to ship this.

File: libs/utils/utils.js [hunk 8]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
+ *
+ * Taxonomy Utils
+ *
+*/
+
+let taxonomyModule;
+
+export function getTaxonomyModule() {
+  return taxonomyModule;
+}
```
Comment (auniverseaway) [inline libs/utils/utils.js:549]: I don't think this provides much value.

File: libs/utils/utils.js [hunk 9]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
+/**
+ * For the given list of topics, returns the corresponding computed taxonomy:
+ * - category: main topic
+ * - topics: tags as an array
+ * - visibleTopics: list of visible topics, including parents
+ * - allTopics: list of all topics, including parents
+ * @param {Array} topics List of topics
+ * @returns {Object} Taxonomy object
+ */
+function computeTaxonomyFromTopics(taxonomy, topics, path) {
```
Comment (auniverseaway) [inline libs/utils/utils.js:560]: This should be moved out of utils. We only put things in utils if a large majority of a site uses a feature.

File: libs/utils/utils.js [hunk 10]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
+        // eslint-disable-next-line no-console
+        console.warn(`Unknown topic in tags list: ${tag} ${path ? `on page ${path}` : '(current page)'}`);
+      }
+    });
+    return { category, topics, visibleTopics, allTopics };
+  }
+  return { category, topics };
+}
+
+export async function loadTaxonomy() {
```
Comment (auniverseaway) [inline libs/utils/utils.js:595]: Same as above. Not an area of concern for utils.

File: libs/utils/utils.js [hunk 11]
```diff
@@ -568,3 +501,218 @@ export function createIntersectionObserver({ el, callback, once = true, options
...
+
+export async function loadArea(area = document) {
+  const config = getConfig();
+  const isDoc = area === document;
+
+  if (isDoc) {
+    decorateHeader();
+  }
+
+  await loadTaxonomy();
```
Comment (auniverseaway) [inline libs/utils/utils.js:672]: Remove. Also: this would load taxonomy for every fragment as well since it's not being checked in the doc.

**Review by auniverseaway — CHANGES_REQUESTED** [review]
Some small comments. Mostly falling back to: don't change so much of utils for your one-off block.

File: libs/scripts/taxonomy.js [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+/**
+ * Returns the taxonomy object
+ * @param {string} lang Language of the taxonomy
+ * @param {*} url URL to use to load the taxonomy
+ * @returns {object} The taxonomy object
+ */
+export default async (lang, url) => {
+  const prefix = lang ? `/${lang}` : '';
+  const root = `${prefix}/topics`;
+  const target = url || `${root}/taxonomy.json`;
```
Comment (auniverseaway) [inline libs/scripts/taxonomy.js:158]: You should use `config.locale.contentRoot` for this.

Comment (ninoaguilar) [reply libs/scripts/taxonomy.js]: Updated 👍🏽

**Review by auniverseaway — APPROVED** [review]

**Comment by auniverseaway** [comment]
@chrischrischris would you mind reviewing? I'm not terribly worried about the performance since this won't ever be LCP.

I don't expect this port from blog to be perfect since @ninoaguilar inherited a lot of stuff (he did clean up a ton), but I do think we should bare minimum be willing to support the block going forward and be able to answer questions if they arise.

**Review by rgclayton — APPROVED** [review]

**Pull request merged**